cpp/demo/poisson-matrix-free/main.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2022 Igor A. Baratta
+//
+// This file is part of DOLFINx (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+//===============================================================//
+// Matrix Free CG solver (C++)
+// =============================================================//
+//
+// This demo illustrates how to:
+// * Solve a linear partial differential equation using a matrix free CG solver.
+// * Create and apply Dirichlet boundary conditions
+// * Compute errors
+//
+// .. math::
+//    - \nabla^{2} u &= f \quad {\rm in} \ \Omega, \\
+//      u &= u_D \quad {\rm on} \ \Gamma_{D}
+//
+//  Where
+// .. math::
+//    u_D &= 1 + x^2 + 2y^2, \\
+//    f = -6
+//
+
+#include "poisson.h"
+#include <cmath>
+#include <dolfinx.h>
+#include <dolfinx/fem/Constant.h>
+#include <xtensor/xarray.hpp>
+
+using namespace dolfinx;
+
+using T = PetscScalar;
+
+namespace linalg
+{
+/// Compute vector r = alpha*x + y
+/// @param[in, out]  r Result
+/// @param[in]  alpha
+/// @param[in]  x
+/// @param[in]  y
+template <typename T>
+void axpy(la::Vector<T>& r, T alpha, const la::Vector<T>& x,
+          const la::Vector<T>& y)
+{
+  std::transform(x.array().cbegin(),
+                 std::next(x.array().cbegin(), x.map()->size_local()),
+                 y.array().cbegin(), r.mutable_array().begin(),
+                 [alpha](auto x, auto y) { return alpha * x + y; });
+}
+
+/// Solve problem A.x = b using the Conjugate Gradient method
+/// @param[in]  b RHS Vector
+/// @param[in, out]  x Solution Vector
+/// @param[in]  action Function that provides the action of the linear operator
+/// @param[in]  kmax Maxmimum number of iterations
+/// @param[in]  rtol Relative tolerances for convergence.
+template <typename T, typename ApplyFunction>
```

Comment:
Maybe use `@tparam` to document `T` and `ApplyFunction`. (`T` is pretty obvious, but `ApplyFunction` less so.)